src/ocean/core/Option.d

```diff
@@ -0,0 +1,236 @@
+
+    public void visit ( void delegate() cb_undefined,
+        void delegate(ref T) cb_defined )
+    {
+        if (this.defined)
+            cb_defined(this.value);
+        else
+            cb_undefined();
+    }
+
+    /**************************************************************************
+
+        A more "old-school" version of `visit`. Provided both to conform style
+        of existing code and avoid delegate bugs in dmd1.
+
+        Discouraged by default as more error-prone than `visit`.
+
+        Parameters:
+            value = out parameter, will be written to with content of `this` if
+            defined
+
+        Return:
+            `true` if this is defined and `value` was updated.
+
+    **************************************************************************/
+
+    public bool get ( ref T value )
+    {
+        if (this.defined) value = this.value;
+        return this.defined;
+    }
+
+    /**************************************************************************
+
+        Returns:
+            `true` if this is defined.
+
+    **************************************************************************/
+
+    public bool isDefined ( )
+    {
+        return this.defined;
+    }
+}
+
+///
+unittest
+{
+    alias Option!(bool) Maybe;
+
+    Maybe x = true;
+    Maybe y = false;
+    Maybe z = Maybe.undefined;
+
+    x.visit(
+        ()               { test(false); },
+        (ref bool value) { test(value); }
+    );
+
+    y.visit(
+        ()               { test(false); },
+        (ref bool value) { test(!value); }
+    );
+
+    z.visit(
+        ()               { test(true); },
+        (ref bool value) { test(false); }
+    );
+}
+
+/******************************************************************************
+
+    Shortcut on top of Option to created defined value, uses IFTI to reduce
+    the noise
+
+    Parameters:
+        value = value to wrap into Option
+
+    Returns:
+        wrapped value
+
+******************************************************************************/
+
+public Option!(T) option ( T ) ( T value )
+{
+    return Option!(T)(value, true);
+}
+
+///
+unittest
+{
+    Option!(int) foo ( bool x )
+    {
+        if (x)
+            return option(42);
+        else
+            return undefined!(int);
+    }
+
+    foo(true).visit(
+        ()              { test(false); },
+        (ref int value) { test(value == 42); }
+    );
+
+    foo(false).visit(
+        ()              { },
+        (ref int value) { test(false); }
+    );
+}
+
+/******************************************************************************
+
+    Shortcut on top of Option to created undefined value.
+
+    Exists only for syntax consistentcy with `option` helper function.
+
+    Parameters:
+        T = type to pass to Option
+
+    Returns:
+        `Option!(T).undefined`
+
+******************************************************************************/
+
+public Option!(T) undefined ( T ) ( )
```

Comment:
Since there is no type inference (unlike `option`) I don't think this adds much value

Reply:
I was going for plain visual syntax consistency with `option` but fter thinking more about it I agree it is not worth new function.